File: mlfootball/matchday.py
"""Matchday — today's (or tomorrow's) World Cup fixtures, in depth.

For each upcoming/live fixture this assembles a full preview out of data the project
already produces, with no new heavy computation:

  * the frozen Dixon-Coles match prediction              (site/data/forecast.json)
  * both value-ranked best XIs + every squad player's
    club / caps / goals / market value                   (squad_value.json + squads.py)
  * each team's manager                                  (Wikipedia squad wikitext)
  * recent international form coming in                   (martj42 history, pre-tournament)
  * an auto factual 'tournament so far' summary          (openfootball live results feed)

Honest about the gaps. There is no clean, free source for per-player *minutes played* at
this tournament — the live feed carries results and goal events only, not lineups — so the
player detail is career caps/goals + club + market value + any tournament goals, and the XI
is value-ranked (same basis as the Squad-Value piece), not a tactical lineup prediction.

Like forecast.json this is a *live* artifact: re-run it (`python -m mlfootball.matchday`) as
results land to refresh scores, scorers and the 'tournament so far' lines.
"""

from __future__ import annotations

import csv
import datetime as dt
import json
import re
import urllib.request
from pathlib import Path

from mlfootball import forecast, squad_value, squads, travel

ROOT = Path(__file__).resolve().parent.parent
RAW = ROOT / "data" / "raw" / "squads_wikitext.txt"
RESULTS_CSV = ROOT / "data" / "results.csv"
FORECAST_JSON = ROOT / "site" / "data" / "forecast.json"
SQUAD_VALUE_JSON = ROOT / "site" / "data" / "squad_value.json"

# everything is keyed by the model/squad spelling (e.g. "Czech Republic", "Turkey",
# "United States", "Ivory Coast") — forecast.json, squad_value.json, squads.py and the
# martj42 history all already agree on it; only the live feed needs translating.
fold = squad_value.fold


# ── managers (Wikipedia squad wikitext) ───────────────────────────────────────
def managers() -> dict[str, str]:
    """{nation: head-coach name} parsed from the cached 2026 squad wikitext.

    Each nation section carries a `Coach: [[Name]]` line; foreign coaches get a
    `{{#invoke:flag|icon|XXX}}` flag prefix we strip. Same section-splitting as
    `squads.load_players`, so the nation keys line up exactly."""
    t = RAW.read_text(encoding="utf-8")
    heads = [(m.start(), m.group(1).strip())
             for m in re.finditer(r"\n===\s*([^=\n][^=]*?)\s*===\n", t)]
    nations = [(p, n) for p, n in heads if not n.lower().startswith(("group", "note"))]
    bounds = [p for p, _ in nations] + [len(t)]
    out: dict[str, str] = {}
    for i, (p, nat) in enumerate(nations):
        # drop HTML comments first — some Coach lines hide one after "Coach:" or
        # even inside the {{#invoke:flag}} template ("Coach:<!--..--> [[Name]]").
        section = re.sub(r"<!--.*?-->", "", t[p:bounds[i + 1]], flags=re.DOTALL)
        m = re.search(
            r"Coach:\s*(?:\{\{[^}]*\}\}\s*)?\[\[(?:[^\]|]*\|)?([^\]]+)\]\]",
            section)
        if m:
            out[nat] = m.group(1).strip()
    return out


# ── recent international form (martj42 history, pre-tournament) ────────────────
def _results_rows() -> list[dict]:
    with open(RESULTS_CSV, encoding="utf-8") as f:
        return list(csv.DictReader(f))


def recent_form(rows: list[dict], team: str, n: int = 5,
                before: str = "2026-06-11") -> dict:
    """Last `n` internationals before kickoff (form coming into the tournament)."""
    games = []
    for r in rows:
        if r["date"] >= before:
            continue
        h, a = r["home_team"], r["away_team"]
        if team not in (h, a):
            continue
        try:
            hs, as_ = int(r["home_score"]), int(r["away_score"])
        except (ValueError, KeyError):
            continue
        gf, ga = (hs, as_) if team == h else (as_, hs)
        neutral = str(r.get("neutral", "")).upper() == "TRUE"
        games.append({
            "date": r["date"],
            "opp": a if team == h else h,
            "gf": gf, "ga": ga,
            "res": "W" if gf > ga else "D" if gf == ga else "L",
            "venue": "N" if neutral else ("H" if team == h else "A"),
        })
    games.sort(key=lambda g: g["date"])
    last = games[-n:]
    return {
        "games": last,
        "w": sum(g["res"] == "W" for g in last),
        "d": sum(g["res"] == "D" for g in last),
        "l": sum(g["res"] == "L" for g in last),
    }


# ── tournament so far (openfootball live feed) ────────────────────────────────
def _ordinal(i: int) -> str:
    return f"{i}{'th' if 11 <= i % 100 <= 13 else {1: 'st', 2: 'nd', 3: 'rd'}.get(i % 10, 'th')}"


def _fetch_feed() -> dict:
    try:
        with urllib.request.urlopen(forecast.FEED_URL, timeout=30) as r:
            return json.loads(r.read().decode("utf-8"))
    except Exception as e:  # noqa: BLE001 — a feed hiccup must not break the build
        print(f"[matchday] feed unavailable ({e}); building with predictions only.")
        return {"matches": []}


def _feed_team(name: str) -> str:
    return forecast.FEED_ALIASES.get(str(name), str(name))


def _ft(m: dict):
    """(home_goals, away_goals) if the match is played, else None."""
    s1, s2 = m.get("score1"), m.get("score2")
    if s1 is None or s2 is None:
        ft = (m.get("score") or {}).get("ft")
        if not ft:
            return None
        s1, s2 = ft[0], ft[1]
    return int(s1), int(s2)


def feed_index(feed: dict) -> dict[tuple, dict]:
    """{(date, frozenset({home,away})): {kickoff, ground, score, scorers}} keyed by model names."""
    idx = {}
    for m in feed.get("matches", []):
        h, a = _feed_team(m.get("team1", "")), _feed_team(m.get("team2", ""))
        if not h or not a:
            continue
        ft = _ft(m)
        idx[(m.get("date", ""), frozenset((h, a)))] = {
            "home": h, "away": a, "kickoff": m.get("time", ""), "ground": m.get("ground", ""),
            "group": (m.get("group", "") or "").replace("Group ", ""),
            "score": {"h": ft[0], "a": ft[1]} if ft else None,
            "scorers": {
                "home": [{"name": g.get("name", ""), "minute": str(g.get("minute", ""))}
                         for g in (m.get("goals1") or [])],
                "away": [{"name": g.get("name", ""), "minute": str(g.get("minute", ""))}
                         for g in (m.get("goals2") or [])],
            },
        }
    return idx


def tournament_so_far(feed: dict) -> dict[str, dict]:
    """Per-team standing + scorers + an auto factual summary, computed from played games."""
    stand: dict[str, dict] = {}
    scorers: dict[str, list] = {}
    for m in feed.get("matches", []):
        ft = _ft(m)
        if not ft:
            continue
        h, a = _feed_team(m.get("team1", "")), _feed_team(m.get("team2", ""))
        gh, ga = ft
        grp = (m.get("group", "") or "").replace("Group ", "")
        for t, gf, gaa in ((h, gh, ga), (a, ga, gh)):
            s = stand.setdefault(t, {"group": grp, "p": 0, "w": 0, "d": 0, "l": 0,
                                     "gf": 0, "ga": 0, "pts": 0})
            s["p"] += 1
            s["gf"] += gf
            s["ga"] += gaa
            if gf > gaa:
                s["w"] += 1
                s["pts"] += 3
            elif gf == gaa:
                s["d"] += 1
                s["pts"] += 1
            else:
                s["l"] += 1
        for t, goals in ((h, m.get("goals1")), (a, m.get("goals2"))):
            for g in (goals or []):
                scorers.setdefault(t, []).append(
                    {"name": g.get("name", ""), "minute": str(g.get("minute", ""))})
    # group ranks
    by_group: dict[str, list] = {}
    for t, s in stand.items():
        by_group.setdefault(s["group"], []).append((t, s))
    rank = {}
    for grp, teams in by_group.items():
        teams.sort(key=lambda ts: (-ts[1]["pts"], -(ts[1]["gf"] - ts[1]["ga"]), -ts[1]["gf"]))
        for i, (t, _) in enumerate(teams):
            rank[t] = i + 1
    out = {}
    for t, s in stand.items():
        gd = s["gf"] - s["ga"]
        wdl = f"W{s['w']} D{s['d']} L{s['l']}"
        summary = (f"P{s['p']} · {wdl} · {s['pts']} pt{'s' if s['pts'] != 1 else ''}"
                   f" · {_ordinal(rank[t])} in Group {s['group']}")
        out[t] = {**s, "gd": gd, "rank": rank[t], "summary": summary,
                  "scorers": sorted(scorers.get(t, []),
                                    key=lambda g: int(re.sub(r"\D", "", g["minute"]) or 0))}
    return out


# ── assemble ──────────────────────────────────────────────────────────────────
def build_export() -> dict:
    fc = json.loads(FORECAST_JSON.read_text())
    sv = json.loads(SQUAD_VALUE_JSON.read_text())
    rows = _results_rows()
    mgr = managers()
    feed = _fetch_feed()
    fidx = feed_index(feed)
    tsf = tournament_so_far(feed)

    # goals per player (squad_value player records carry caps but not goals)
    goals_by = {}
    for p in squads.load_players():
        goals_by.setdefault(p["nation"], {})[fold(p["name"])] = p["goals"]

    teams = {}
    for nat, info in sv["nations"].items():
        gmap = goals_by.get(nat, {})
        primary, secondary = squads.FED_COLORS.get(nat, ("#5B5751", "#8A857C"))
        players = [{
            "name": p["name"], "pos": p["pos"], "pos_group": p["pos_group"],
            "club": p["club"], "caps": p["caps"], "goals": gmap.get(fold(p["name"])),
            "value_eur": p["value_eur"], "in_xi": p["in_xi"],
        } for p in info["players"]]
        teams[nat] = {
            "flag": info["flag"], "colors": {"primary": primary, "secondary": secondary},
            "manager": mgr.get(nat),
            "squad_value": info["squad_value"], "xi_value": info["xi_value"],
            "n_players": info["n_players"], "n_valued": info["n_valued"],
            "recent_form": recent_form(rows, nat),
            "tournament": tsf.get(nat),
            "players": players,
        }

    fixtures = []
    for f in fc["fixtures"]:
        key = (f["date"], frozenset((f["home"], f["away"])))
        live = fidx.get(key, {})
        fixtures.append({
            "date": f["date"], "group": f["group"], "home": f["home"], "away": f["away"],
            "venue": f.get("venue", ""), "kickoff": live.get("kickoff", ""),
            "ground": live.get("ground", ""),
            "status": "done" if live.get("score") else "upcoming",
            "score": live.get("score"), "scorers": live.get("scorers"),
            "model": {
                "p_home": f.get("p_home"), "p_draw": f.get("p_draw"), "p_away": f.get("p_away"),
                "exp_home": f.get("exp_home"), "exp_away": f.get("exp_away"),
                "top_score": (f.get("top_scores") or [None])[0],
                "known": f.get("known", False),
            },
        })
    fixtures.sort(key=lambda f: (f["date"], f["group"], f["home"]))

    today = dt.date.today().isoformat()
    return {
        "meta": {
            "generated_for": today,
            "generated_at": dt.datetime.now(dt.timezone.utc).isoformat(timespec="seconds"),
            "asof_model": fc["meta"]["asof"],
            "n_fixtures": len(fixtures),
            "n_played": sum(f["status"] == "done" for f in fixtures),
            "note": ("XIs are value-ranked (Transfermarkt), not tactical lineups; the live feed "
                     "carries results and goal events only, so there are no per-player minutes."),
        },
        "teams": teams,
        "fixtures": fixtures,
        "sources": [
            {"label": "Predictions — Dixon-Coles model (this project)", "url": ""},
            {"label": "Squads & managers — Wikipedia 2026 World Cup squads",
             "url": "https://en.wikipedia.org/wiki/2026_FIFA_World_Cup_squads"},
            {"label": "Market values — Transfermarkt", "url": "https://www.transfermarkt.com"},
            {"label": "Live results — openfootball/worldcup.json",
             "url": "https://github.com/openfootball/worldcup.json"},
            {"label": "Match history — martj42 international results",
             "url": "https://github.com/martj42/international_results"},
        ],
    }


def export(path: Path | None = None) -> dict:
    out = build_export()
    path = path or ROOT / "site" / "data" / "matchday.json"
    path.write_text(json.dumps(out, indent=2, ensure_ascii=False))
    return out


if __name__ == "__main__":
    d = export()
    m = d["meta"]
    print(f"matchday.json: {m['n_fixtures']} fixtures ({m['n_played']} played), "
          f"{len(d['teams'])} teams, generated_for {m['generated_for']}")
    today = [f for f in d["fixtures"] if f["date"] == m["generated_for"]]
    print(f"  today ({m['generated_for']}): " +
          (", ".join(f"{f['home']} v {f['away']}" for f in today) or "no fixtures"))
    miss_mgr = [n for n, t in d["teams"].items() if not t["manager"]]
    if miss_mgr:
        print(f"  ! no manager parsed for: {miss_mgr}")

File: site/js/matchday.js
/* Matchday — hub piece 09. One static matchday.json: today's (or the next) fixtures,
   each opened into a full preview — model call, both value-ranked XIs on a pitch,
   managers, form, and tournament-so-far. Vanilla JS, hub palette, no build step. */
(function () {
  "use strict";

  var DATA = null, dates = [], windowDate = null, sel = null;
  var WEEKDAY = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
  var MONTH = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
  var POS_ORDER = ["Goalkeepers", "Defenders", "Midfielders", "Forwards"];
  var POS_SHORT = { Goalkeepers: "GK", Defenders: "DF", Midfielders: "MF", Forwards: "FW" };

  function pct(x) { return x == null || isNaN(x) ? "—" : Math.round(100 * x) + "%"; }
  function eur(v) {
    if (v == null || !v) return "—";
    return v >= 1e9 ? "€" + (v / 1e9).toFixed(2) + "bn" : "€" + Math.round(v / 1e6) + "m";
  }
  function el(tag, cls, html) {
    var e = document.createElement(tag);
    if (cls) e.className = cls;
    if (html != null) e.innerHTML = html;
    return e;
  }
  function esc(s) { return String(s == null ? "" : s).replace(/[&<>"]/g, function (c) {
    return { "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" }[c]; }); }
  function lastName(n) { var p = String(n || "").split(" "); return p[p.length - 1]; }
  function fmtDate(iso) {
    var d = new Date(iso + "T00:00:00Z");
    return WEEKDAY[d.getUTCDay()] + " " + d.getUTCDate() + " " + MONTH[d.getUTCMonth()];
  }
  function todayISO() {
    var n = new Date();
    return n.getFullYear() + "-" + String(n.getMonth() + 1).padStart(2, "0") + "-" +
      String(n.getDate()).padStart(2, "0");
  }
  function isUpcoming(f) { return f.status !== "done"; }
  function team(name) { return (DATA.teams || {})[name] || {}; }

  // ── boot ────────────────────────────────────────────────────────────────────
  fetch("../data/matchday.json").then(function (r) { return r.json(); })
    .then(init)
    .catch(function () {
      document.getElementById("md-preview").innerHTML =
        '<p class="md-empty">Couldn\'t load the matchday data.</p>';
    });

  function init(d) {
    DATA = d;
    dates = Array.from(new Set(d.fixtures.map(function (f) { return f.date; }))).sort();
    fillStats();
    windowDate = computeWindow();
    bindNav();
    document.getElementById("md-methodtext").innerHTML =
      "The model is the project's Dixon-Coles goals model, <b>frozen the day before the tournament</b> " +
      "(" + esc(d.meta.asof_model) + ") — so every match call here is a genuine pre-tournament prediction, " +
      "the same engine behind the Forecast piece. The two elevens are <b>value-ranked</b> (top goalkeeper " +
      "plus the ten most valuable outfielders, from Transfermarkt) — a strength proxy, <b>not a tactical " +
      "lineup prediction</b>, so a fixture's true XI may differ. Form is each side's last five internationals " +
      "before kickoff (martj42 history); the tournament line is computed live from results. " +
      "<b>What's missing, honestly:</b> there is no clean, free source for per-player minutes at this " +
      "World Cup — the live feed carries results and goal events only, not lineups — so player detail is " +
      "career caps, goals, club and market value, plus any goals scored at this tournament.";
    selectDate(windowDate);
  }

  function fillStats() {
    var s = document.getElementById("stats");
    var inWin = DATA.fixtures.filter(function (f) { return f.date === computeWindow(); }).length;
    [{ n: inWin, l: "games on the card" },
     { n: DATA.meta.n_played + "<small>/" + DATA.meta.n_fixtures + "</small>", l: "group games played" },
     { n: 48, l: "squads, fully priced" }
    ].forEach(function (c) {
      var d = el("div", "stat");
      d.appendChild(el("div", "num", c.n));
      d.appendChild(el("div", "lbl", c.l));
      s.appendChild(d);
    });
  }

  // pick the matchday to open on: today if it has an unplayed game, else the next
  // date with fixtures; clamp to the tournament window for out-of-range visitors.
  function computeWindow() {
    var ref = todayISO();
    if (ref < dates[0]) return dates[0];
    if (ref > dates[dates.length - 1]) return dates[dates.length - 1];
    var onRef = DATA.fixtures.filter(function (f) { return f.date === ref; });
    if (onRef.length && onRef.some(isUpcoming)) return ref;
    var nxt = dates.filter(function (dd) { return dd > ref; })[0];
    return nxt || dates[dates.length - 1];
  }

  function bindNav() {
    document.getElementById("md-prev").onclick = function () { stepDay(-1); };
    document.getElementById("md-next").onclick = function () { stepDay(1); };
  }
  function stepDay(dir) {
    var i = dates.indexOf(windowDate) + dir;
    if (i < 0 || i >= dates.length) return;
    selectDate(dates[i]);
  }

  function whenLabel(date) {
    var ref = todayISO();
    var di = dates.indexOf(date);
    if (date === ref) return "Today";
    // tomorrow relative to real today
    var t = new Date(ref + "T00:00:00Z"); t.setUTCDate(t.getUTCDate() + 1);
    var tomorrow = t.toISOString().slice(0, 10);
    if (date === tomorrow) return "Tomorrow";
    return di === 0 ? "Opening day" : di === dates.length - 1 ? "Final group day" : "Matchday";
  }

  function selectDate(date) {
    windowDate = date;
    document.getElementById("md-prev").disabled = dates.indexOf(date) <= 0;
    document.getElementById("md-next").disabled = dates.indexOf(date) >= dates.length - 1;
    document.getElementById("md-when").textContent = whenLabel(date);
    document.getElementById("md-date").textContent = fmtDate(date);
    var fx = DATA.fixtures.filter(function (f) { return f.date === date; })
      .sort(function (a, b) { return (a.kickoff || "").localeCompare(b.kickoff || "") || a.group.localeCompare(b.group); });
    renderCards(fx);
    var def = fx.filter(isUpcoming)[0] || fx[0];
    if (def) selectFixture(def, fx);
    else document.getElementById("md-preview").innerHTML =
      '<p class="md-empty">No fixtures on this date.</p>';
  }

  function renderCards(fx) {
    var box = document.getElementById("md-cards"); box.innerHTML = "";
    fx.forEach(function (f) {
      var c = el("div", "md-card");
      c.dataset.key = f.home + "|" + f.away;
      var th = team(f.home), ta = team(f.away);
      c.appendChild(el("div", "teams",
        '<span class="flag">' + (th.flag || "") + "</span> " + esc(f.home) +
        ' <span style="color:var(--muted)">v</span> ' + esc(f.away) +
        ' <span class="flag">' + (ta.flag || "") + "</span>"));
      var right = f.status === "done" && f.score
        ? '<span class="sc">' + f.score.h + "–" + f.score.a + "</span>"
        : (f.model && f.model.p_home != null ? favLabel(f) : "");
      c.appendChild(el("div", "meta",
        "<span>" + esc(f.kickoff || ("Grp " + f.group)) + "</span><span>" + right + "</span>"));
      c.onclick = function () { selectFixture(f, fx); };
      box.appendChild(c);
    });
  }
  function favLabel(f) {
    var m = f.model, best = Math.max(m.p_home, m.p_draw, m.p_away);
    var who = best === m.p_home ? f.home : best === m.p_away ? f.away : "draw";
    return esc(who === "draw" ? "draw" : lastName(who)) + " " + pct(best);
  }

  function selectFixture(f, fx) {
    sel = f;
    Array.prototype.forEach.call(document.querySelectorAll(".md-card"), function (c) {
      c.classList.toggle("sel", c.dataset.key === f.home + "|" + f.away);
    });
    renderPreview(f);
  }

  // ── the preview ───────────────────────────────────────────────────────────────
  function renderPreview(f) {
    var box = document.getElementById("md-preview"); box.innerHTML = "";
    var th = team(f.home), ta = team(f.away);

    // header
    var head = el("div", "md-head");
    head.appendChild(teamHead(f.home, th, false));
    var center = el("div", "md-center");
    if (f.status === "done" && f.score) {
      center.innerHTML = '<div class="ko">' + esc(f.kickoff || "") + " · Grp " + f.group + "</div>" +
        '<div class="score">' + f.score.h + "–" + f.score.a + "</div>" +
        '<span class="md-badge done">full time</span>';
    } else {
      center.innerHTML = '<div class="ko">' + esc(f.kickoff || "") + " · Grp " + f.group + "</div>" +
        '<div class="vs">v</div><span class="md-badge upcoming">upcoming</span>';
    }
    head.appendChild(center);
    head.appendChild(teamHead(f.away, ta, true));
    box.appendChild(head);
    box.appendChild(el("p", "md-venue", esc(f.venue || f.ground || "")));

    if (f.status === "done" && f.scorers) box.appendChild(scorerLine(f));

    // model strip
    if (f.model && f.model.p_home != null) box.appendChild(modelStrip(f));

    // dossiers
    var dos = el("div", "md-dossier");
    dos.appendChild(dossier(f.home, th));
    dos.appendChild(dossier(f.away, ta));
    box.appendChild(dos);

    // pitch
    box.appendChild(pitchPanel(f, th, ta));

    // full squads
    var sq = el("div", "md-squads");
    sq.appendChild(squadList(f.home, th));
    sq.appendChild(squadList(f.away, ta));
    box.appendChild(sq);
  }

  function teamHead(name, t, away) {
    var d = el("div", "md-team" + (away ? " away" : ""));
    d.appendChild(el("span", "flag", t.flag || ""));
    var box = el("div");
    box.appendChild(el("div", "nm", esc(name)));
    box.appendChild(el("div", "mgr", t.manager ? "Mgr · <b>" + esc(t.manager) + "</b>" : ""));
    d.appendChild(box);
    return d;
  }

  function scorerLine(f) {
    function side(list, who) {
      if (!list || !list.length) return "";
      return "<b>" + esc(who) + ":</b> " + list.map(function (g) {
        return esc(g.name) + (g.minute ? " " + esc(g.minute) + "'" : "");
      }).join(", ");
    }
    var parts = [side(f.scorers.home, f.home), side(f.scorers.away, f.away)].filter(Boolean);
    return el("p", "md-scorers", parts.length
      ? '<div style="text-align:center">⚽ ' + parts.join(" &nbsp;·&nbsp; ") + "</div>" : "");
  }

  function modelStrip(f) {
    var m = f.model, wrap = el("div", "md-model");
    wrap.appendChild(el("h4", null, "Model call — frozen before kickoff"));
    var bar = el("div", "md-bar");
    [["h", m.p_home], ["d", m.p_draw], ["a", m.p_away]].forEach(function (seg) {
      var w = (100 * seg[1]).toFixed(1) + "%";
      var d = el("div", seg[0], (seg[1] >= 0.1 ? pct(seg[1]) : ""));
      d.style.width = w; bar.appendChild(d);
    });
    wrap.appendChild(bar);
    wrap.appendChild(el("div", "md-barlbl",
      "<span>" + esc(f.home) + " win</span><span>draw</span><span>" + esc(f.away) + " win</span>"));
    var xg = el("div", "md-xg");
    xg.innerHTML = "expected goals &nbsp; <b>" + (m.exp_home != null ? m.exp_home.toFixed(1) : "—") +
      "</b> &nbsp;–&nbsp; <b>" + (m.exp_away != null ? m.exp_away.toFixed(1) : "—") + "</b>" +
      (m.top_score ? '<span style="margin-left:18px">most likely &nbsp;<b>' +
        m.top_score.h + "–" + m.top_score.a + "</b> (" + pct(m.top_score.p) + ")</span>" : "");
    wrap.appendChild(xg);
    return wrap;
  }

  function dossier(name, t) {
    var d = el("div", "md-dos");
    d.appendChild(el("div", "dos-team", '<span style="font-size:18px">' + (t.flag || "") +
      "</span> " + esc(name)));
    // form
    var rf = t.recent_form || { games: [], w: 0, d: 0, l: 0 };
    var chips = rf.games.map(function (g) {
      return '<span class="fchip ' + g.res + '" title="' + esc(g.date) + " v " + esc(g.opp) +
        " " + g.gf + "-" + g.ga + '">' + g.res + "</span>";
    }).join("");
    var formRow = el("div", "dos-row");
    formRow.innerHTML = '<span class="k">Form coming in</span>' +
      '<div class="md-form">' + (chips || '<span class="ftext">no recent matches</span>') +
      '<span class="ftext">W' + rf.w + " D" + rf.d + " L" + rf.l + "</span></div>";
    var last = rf.games.slice(-3).reverse().map(function (g) {
      return (g.res === "W" ? "✓ " : g.res === "L" ? "✗ " : "= ") + g.gf + "–" + g.ga + " v " + esc(g.opp);
    }).join(" · ");
    if (last) formRow.appendChild(el("div", "md-results", last));
    d.appendChild(formRow);
    // tournament so far
    var tr = el("div", "dos-row");
    if (t.tournament) {
      tr.innerHTML = '<span class="k">This tournament</span><div class="md-tour"><span class="summ">' +
        esc(t.tournament.summary) + "</span></div>";
      if (t.tournament.scorers && t.tournament.scorers.length)
        tr.appendChild(el("div", "md-scorers", "Scorers: " + t.tournament.scorers.map(function (g) {
          return esc(g.name) + (g.minute ? " " + esc(g.minute) + "'" : "");
        }).join(", ")));
    } else {
      tr.innerHTML = '<span class="k">This tournament</span><div class="md-tour" style="color:var(--muted)">Yet to play.</div>';
    }
    d.appendChild(tr);
    // value
    d.appendChild(el("div", "dos-row",
      '<span class="k">Squad value · best XI</span>' + eur(t.squad_value) + " · " + eur(t.xi_value) +
      ' <span style="color:var(--muted)">(' + (t.n_valued || 0) + "/" + (t.n_players || 0) + " priced)</span>"));
    return d;
  }

  // ── the pitch (both value-ranked XIs facing off) ──────────────────────────────
  function xiByLine(t) {
    var xi = (t.players || []).filter(function (p) { return p.in_xi; });
    var lines = { Goalkeepers: [], Defenders: [], Midfielders: [], Forwards: [] };
    xi.forEach(function (p) { (lines[p.pos_group] || lines.Midfielders).push(p); });
    return lines;
  }
  function pitchPanel(f, th, ta) {
    var wrap = el("div", "md-pitchwrap");
    wrap.appendChild(el("div", "md-section-h", "Likely elevens — value-ranked"));
    var W = 1000, H = 560;
    var svg = '<svg class="md-pitch" viewBox="0 0 ' + W + " " + H + '" role="img" aria-label="Pitch with both value-ranked elevens">';
    // turf + markings
    svg += '<rect x="0" y="0" width="' + W + '" height="' + H + '" rx="14" fill="#eef3ec"/>';
    svg += '<rect x="14" y="14" width="' + (W - 28) + '" height="' + (H - 28) + '" rx="8" fill="none" stroke="#cdd9c8" stroke-width="2"/>';
    svg += '<line x1="' + (W / 2) + '" y1="14" x2="' + (W / 2) + '" y2="' + (H - 14) + '" stroke="#cdd9c8" stroke-width="2"/>';
    svg += '<circle cx="' + (W / 2) + '" cy="' + (H / 2) + '" r="60" fill="none" stroke="#cdd9c8" stroke-width="2"/>';
    svg += '<rect x="14" y="' + (H / 2 - 110) + '" width="120" height="220" fill="none" stroke="#cdd9c8" stroke-width="2"/>';
    svg += '<rect x="' + (W - 134) + '" y="' + (H / 2 - 110) + '" width="120" height="220" fill="none" stroke="#cdd9c8" stroke-width="2"/>';
    svg += sidePlayers(xiByLine(th), false, W, H, "var(--amber)");
    svg += sidePlayers(xiByLine(ta), true, W, H, "var(--accent)");
    svg += "</svg>";
    wrap.appendChild(el("div", null, svg));

    // value mismatch bar
    var hv = th.xi_value || 0, av = ta.xi_value || 0, tot = hv + av || 1;
    var vm = el("div", "md-vmis");
    vm.innerHTML = '<div class="md-vbar"><div class="h" style="width:' + (100 * hv / tot) +
      '%"></div><div class="a" style="width:' + (100 * av / tot) + '%"></div></div>' +
      '<div class="md-vlbl"><span><b>' + eur(hv) + "</b> " + esc(f.home) + " XI</span>" +
      "<span>" + esc(f.away) + " XI <b>" + eur(av) + "</b></span></div>";
    wrap.appendChild(vm);
    return wrap;
  }
  function sidePlayers(lines, away, W, H, color) {
    // column x positions from each team's own goal toward the centre
    var cols = away ? { Goalkeepers: .955, Defenders: .80, Midfielders: .655, Forwards: .53 }
                    : { Goalkeepers: .045, Defenders: .20, Midfielders: .345, Forwards: .47 };
    var out = "";
    POS_ORDER.forEach(function (grp) {
      var ps = lines[grp] || [], n = ps.length;
      ps.forEach(function (p, i) {
        var cx = cols[grp] * W;
        var cy = (H - 36) * (i + 1) / (n + 1) + 18;
        var nm = lastName(p.name);
        out += '<g><circle cx="' + cx.toFixed(0) + '" cy="' + cy.toFixed(0) + '" r="15" fill="' +
          color + '" stroke="#fff" stroke-width="2"><title>' + esc(p.name) + " — " + esc(p.pos) +
          ", " + esc(p.club) + " · " + (p.caps || 0) + " caps" +
          (p.goals != null ? ", " + p.goals + " gls" : "") + " · " + eur(p.value_eur) + "</title></circle>";
        out += '<text class="md-pl-name" x="' + cx.toFixed(0) + '" y="' + (cy - 22).toFixed(0) +
          '" text-anchor="middle" font-size="15" fill="#1a1714">' + esc(nm) + "</text>";
        out += '<text class="md-pl-val" x="' + cx.toFixed(0) + '" y="' + (cy + 30).toFixed(0) +
          '" text-anchor="middle" font-size="12" fill="#7a6e63">' + eur(p.value_eur) + "</text></g>";
      });
    });
    return out;
  }

  function squadList(name, t) {
    var d = el("div", "md-sq");
    var players = (t.players || []);
    d.appendChild(el("h4", null, '<span>' + (t.flag || "") + " " + esc(name) +
      "</span><b>" + (t.n_players || players.length) + "</b>"));
    // XI first (value order already), then bench
    var xi = players.filter(function (p) { return p.in_xi; });
    var bench = players.filter(function (p) { return !p.in_xi; });
    xi.concat(bench).forEach(function (p) {
      var row = el("div", "md-pl " + (p.in_xi ? "xi" : "bench"));
      row.innerHTML = '<span class="pp">' + (POS_SHORT[p.pos_group] || p.pos) + "</span>" +
        '<span class="pn">' + esc(p.name) + " <small>" + esc(p.club) +
        " · " + (p.caps || 0) + "c" + (p.goals != null ? "/" + p.goals + "g" : "") + "</small></span>" +
        '<span class="pv">' + eur(p.value_eur) + "</span>";
      d.appendChild(row);
    });
    return d;
  }
})();
